Update Conan dependencies: protobuf and grpc

bthomee opened the pull request:
## High Level Overview of Change

This PR updates protobuf and grpc to their latest versions.

### Context of Change

Updating dependencies is now possible due to the switch to Conan 2. The latest protobuf version no longer requires patches, so we can use it directly from the official Conan Center Index, while the latest grpc still needed a patch, which was added to our own Conan Center Index fork in https://github.com/XRPLF/conan-center-index/pull/8.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

bthomee committed:
Updates protobuf and grpc

bthomee committed:
Explicitly specify version of abseil to use

bthomee committed:
Temporarily disable ripple remote

bthomee committed:
Temporarily disable ripple remote

bthomee committed:
Temporarily disable ripple remote

bthomee committed:
Temporarily compile using Clang 17

bthomee committed:
Temporarily compile using Clang 17

bthomee committed:
Restore currently used image versions

Bronek commented:
This is failing because of portability bugs in `protobuf` which were recently fixed:

* [Unable to build v30.0 on debian-12 with clang-16 #20645 (protobuf)](https://github.com/protocolbuffers/protobuf/issues/20645)
* [Unable to build v31.0 with clang-cl in c++20 debug build #21957 (protobuf)](https://github.com/protocolbuffers/protobuf/issues/21957)

I raised a PR in conan-center-index to backport both fixes to `protobuf/6.30.1` Conan package

* [Fix protobuf/6.30.1 compatibility issues with clang-16 and MSVC #27982 (conan-center-index)](https://github.com/conan-io/conan-center-index/pull/27982)

If this PR fails, we will probably need to add an equivalent `external/protobuf`.

Bronek commented:
The Windows errors are because of conflict of runtime in `grpc` plugins, for example:

```
libprotoc.lib(code_generator.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in ruby_plugin.obj [C:\Users\runneradmin\.conan2\p\b\grpc64c32e704d30a\b\build\grpc_ruby_plugin.vcxproj]
```

Looking at conan source for `grpc` , there's lots of plugins and most of them enabled by default. I think we can safely disable most (all ?) of them

https://github.com/conan-io/conan-center-index/blob/master/recipes/grpc/all/conanfile.py

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Adds protobuf patches

bthomee committed:
Adds protobuf patches

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee commented:
> The Windows errors are because of conflict of runtime in `grpc` plugins, for example:
> 
> ```
> libprotoc.lib(code_generator.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MD_DynamicRelease' in ruby_plugin.obj [C:\Users\runneradmin\.conan2\p\b\grpc64c32e704d30a\b\build\grpc_ruby_plugin.vcxproj]
> ```
> 
> Looking at conan source for `grpc` , there's lots of plugins and most of them enabled by default. I think we can safely disable most (all ?) of them
> 
> https://github.com/conan-io/conan-center-index/blob/master/recipes/grpc/all/conanfile.py

I set the following options:
```
'grpc/*:shared': False,
'grpc/*:secure': True,
'grpc/*:codegen': False,
'grpc/*:cpp_plugin': True,
'grpc/*:csharp_ext': False,
'grpc/*:csharp_plugin': False,
'grpc/*:node_plugin': False,
'grpc/*:objective_c_plugin': False,
'grpc/*:php_plugin': False,
'grpc/*:python_plugin': False,
'grpc/*:ruby_plugin': False,
'grpc/*:otel_plugin': False,
```

essentially explicitly disabling all these plugins except the C++ one, yet then got the failure:
```
CMake Error at /root/.conan2/p/b/protoa39c0becc72f6/p/lib/cmake/protobuf/protobuf-generate.cmake:149 (add_custom_command):
  Error evaluating generator expression:

    $<TARGET_FILE:gRPC::grpc_cpp_plugin>

  No target "gRPC::grpc_cpp_plugin"
Call Stack (most recent call first):
  cmake/target_protobuf_sources.cmake:42 (protobuf_generate)
  cmake/RippledCore.cmake:25 (target_protobuf_sources)
  CMakeLists.txt:149 (include)
```

This needs a closer look.

bthomee committed:
Enable codegen grpc plugin

bthomee committed:
Enable codegen grpc plugin

bthomee commented:
> This needs a closer look.

After setting `'grpc/*:codegen': True,` compilation succeeds again. Only Windows now needs attention.

bthomee committed:
Merge branch 'develop' into bthomee/proto

ximinez commented:
Re: The Windows CI build failures, I built this successfully on my machine with no modifications.

bthomee committed:
Switches Windows to use dynamic runtime

Bronek committed:
Build options cleanup (#5581)

As we no longer support old compiler versions, we are bringing back some warnings by removing no longer relevant `-Wno-...` options.

mathbunnyru commented:
I didn't know the answer, but I found out!

So, `protobuf` recipe already has the code like this:
https://github.com/conan-io/conan-center-index/blob/master/recipes/protobuf/all/conanfile.py#L153-L156

I think it makes sense to add 100% identical code to grpc (changing the variable name).
I still highly suggest to create a PR in the CCI repository.

mathbunnyru committed:
ci: Build all conan dependencies from source for now (#5623)

shawnxie999 committed:
fix `DeliveredAmount` and `delivered_amount` in transaction metadata for direct MPT transfer (#5569)

The Payment transaction metadata is missing the `DeliveredAmount` field that displays the actual amount delivered to the destination excluding transfer fees. This amendment fixes this problem.

Bronek committed:
Fix crash in Slot::deletePeer (#5635)

Fix crash due to recurrent call to `Slot::deletePeer` (via `OverlayImpl::unsquelch`) when a peer is disconnected at just the wrong moment.

mathbunnyru committed:
Revert "ci: Build all conan dependencies from source for now (#5623)" (#5639)

This reverts commit 9b45b6888b9bfc4ca81d0fb71e0b6f2f3a120bae.

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Remove external libraries hosted in our conan center index fork

bthomee committed:
Update Conan remote

bthomee committed:
Remove unnecessary authentication for read-only remote access, and update remote name

bthomee committed:
Remove unnecessary section in readme

bthomee commented:
I propose we merge https://github.com/XRPLF/rippled/pull/5643 first, which removes the `external/` directories completely and removes logging into the Conan remote as that isn't required for read-only access. Once those changes have been made, this PR becomes much simpler.

bthomee commented on BUILD.md:
````diff
@@ -217,19 +228,6 @@ It patches their CMake to correctly import its dependencies.
     `--settings build_type=$BUILD_TYPE` or in the profile itself,
     under the section `[settings]` with the key `build_type`.
 
-    If you are using a Microsoft Visual C++ compiler,
-    then you will need to ensure consistency between the `build_type` setting
-    and the `compiler.runtime` setting.
-
-    When `build_type` is `Release`, `compiler.runtime` should be `MT`.
-
-    When `build_type` is `Debug`, `compiler.runtime` should be `MTd`.
-
-    ```
-    conan install .. --output-folder . --build missing --settings build_type=Release --settings compiler.runtime=MT
-    conan install .. --output-folder . --build missing --settings build_type=Debug --settings compiler.runtime=MTd
-    ```
````
These comments are unnecessary. One only needs to specify `dynamic` for the `compiler.runtime` setting rather than the separate `MD` / `MDd` values (when using the dynamic runtime, as proposed in this PR; for static runtime it would be `MT` and `MTd`). Moreover, our default profile already sets the compiler runtime appropriately, and thus developers do not need to do anything else.

bthomee committed:
Merge branch 'develop' into bthomee/external

bthomee committed:
Merge branch 'develop' into bthomee/external

bthomee committed:
Merge branch 'develop' into bthomee/external

bthomee committed:
Add instructions on how to fetch updated recipes from CCI fork

bthomee committed:
Fix link in documentation

bthomee committed:
Pull in latest changes

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Fix conflict

bthomee committed:
Merge branch 'bthomee/external' into bthomee/proto

bthomee committed:
Fix commands in readme

Bronek commented:
Just had a nice conversation with Conan maintainers in https://github.com/conan-io/conan-center-index/pull/27982 and turns out, we would do better if we waited for protobuf 6.32.x - which currently is being prepared for release in Conan Center.

See this comment and below https://github.com/conan-io/conan-center-index/pull/27982#issuecomment-3158884624

This also means that we should **not** put a patched 6.30.1 in our own Conan

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee commented:
I brought the branch up to date with `develop` but will wait for `protobuf` v6.32.x - nothing yet in the Conan Center, unfortunately.

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Use updated protobuf 6.32.1

bthomee committed:
Update conan.lock with Linux-only dependencies

bthomee committed:
Temporarily use custom RHEL 9 docker image to test missing dependencies

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Override compiler.cppstd for M4

bthomee committed:
Override compiler.cppstd for M4

bthomee committed:
Restore profile, using patched M4 recipe instead

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Update m4

bthomee commented:
@Bronek @ximinez Patching the `m4` recipe worked - all builds succeed.

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Remove Conan profile override for grpc

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Update Conan lock file

bthomee committed:
Update Conan lock file

bthomee committed:
Merge branch 'develop' into bthomee/proto

bthomee committed:
Restore cmake changes

bthomee commented on BUILD.md:
````diff
@@ -383,19 +388,6 @@ tools.build:cxxflags=['-DBOOST_ASIO_DISABLE_CONCEPTS']
    `--settings build_type=$BUILD_TYPE` or in the profile itself,
    under the section `[settings]` with the key `build_type`.
 
-   If you are using a Microsoft Visual C++ compiler,
-   then you will need to ensure consistency between the `build_type` setting
-   and the `compiler.runtime` setting.
-
-   When `build_type` is `Release`, `compiler.runtime` should be `MT`.
-
-   When `build_type` is `Debug`, `compiler.runtime` should be `MTd`.
-
-   ```
-   conan install .. --output-folder . --build missing --settings build_type=Release --settings compiler.runtime=MT
-   conan install .. --output-folder . --build missing --settings build_type=Debug --settings compiler.runtime=MTd
-   ```
-
````
This is no longer relevant, since the `XrplCompiler.cmake` sets these flags automatically on Windows.

vlntb commented on conanfile.py:
```diff
@@ -59,6 +59,16 @@ class Xrpl(ConanFile):
+        "grpc/*:php_plugin": False,
+        "grpc/*:python_plugin": False,
+        "grpc/*:ruby_plugin": False,
+        "grpc/*:otel_plugin": False,
```
What is the need to add plugins to support: NodeJS, ObjectiveC, PHP, Python, Ruby, Otel, etc.?
To my knowledge we don't use any of these tools right now.

bthomee replied:
The Conan recipe sets them to `True` by default (see https://github.com/conan-io/conan-center-index/blob/master/recipes/grpc/all/conanfile.py), so I'm intentionally disabling them by setting them to `False` here.

bthomee committed:
Merge branch 'develop' into bthomee/proto

pratikmankawde commented on BUILD.md:
````diff
@@ -142,20 +142,25 @@ locally:
 
 ```bash
 cd external
+mkdir -p conan-center-index
````
I think `-p` is not needed here since we are only creating the target directory, no parents.

bthomee replied:
If the directory already exists, then the call without `-p` will return an error.

```
$> cd external
$> mkdir conan-center-index
$> mkdir conan-center-index
mkdir: conan-center-index: File exists
$> mkdir -p conan-center-index
[no error]
```

pratikmankawde replied:
Such error should be fine I suppose, since this is not a script rather a doc file.

bthomee replied:
Sure, but I prefer to avoid exposing the developer to unnecessary warnings/errors.

pratikmankawde commented on BUILD.md:
```diff
@@ -142,20 +142,25 @@ locally:
-conan export --version 4.0.3 recipes/soci/all
 rm -rf .git
+cd ../..
+conan export --version 2015.03 external/conan-center-index/recipes/ed25519/all
+conan export --version 1.72.0 external/conan-center-index/recipes/grpc/all
+conan export --version 0.7.0 external/conan-center-index/recipes/secp256k1/all
+conan export --version 1.1.10 external/conan-center-index/recipes/snappy/all
+conan export --version 4.0.3 external/conan-center-index/recipes/soci/all
```
Can we avoid hardcoding versions here? It will minimize maintaining these versions. We can just redirect developers to take versions from the lockfile maybe?

bthomee replied:
I'll take a look if I can come up with a command that extracts the version number from the lock file. However, that will be done in a follow-up PR as it is not critical to the functionality introduced here.

pratikmankawde replied:
I was hinting towards manual work for the developer to pull those version numbers.

bthomee replied:
See https://github.com/XRPLF/rippled/pull/6112. It updates the documentation to automatically pull the version numbers from the `conan.lock` file.

pratikmankawde commented:
Minor comments, looks fine otherwise.

pratikmankawde reviewed (approved):
LGTM, minor suggestions.
